cmd/crypto/kms.go

```diff
@@ -33,74 +33,27 @@ import (
 // associated with a certain object.
 type Context map[string]string
 
-// WriteTo writes the context in a canonical from to w.
-// It returns the number of bytes and the first error
-// encounter during writing to w, if any.
-//
-// WriteTo sorts the context keys and writes the sorted
-// key-value pairs as canonical JSON object to w.
-// Sort order is based on the un-escaped keys.
-//
-// Note that neither keys nor values are escaped for JSON.
-func (c Context) WriteTo(w io.Writer) (n int64, err error) {
-	sortedKeys := make(sort.StringSlice, 0, len(c))
-	for k := range c {
-		sortedKeys = append(sortedKeys, k)
-	}
-	sort.Sort(sortedKeys)
+// MarshalText returns a canonical text representation of
+// the Context.
 
-	escape := func(s string) string {
-		buf := bytes.NewBuffer(make([]byte, 0, len(s)))
-		EscapeStringJSON(buf, s)
-		return buf.String()
-	}
-
-	nn, err := io.WriteString(w, "{")
-	if err != nil {
-		return n + int64(nn), err
-	}
-	n += int64(nn)
-	for i, k := range sortedKeys {
-		s := fmt.Sprintf("\"%s\":\"%s\",", escape(k), escape(c[k]))
-		if i == len(sortedKeys)-1 {
-			s = s[:len(s)-1] // remove last ','
-		}
-
-		nn, err = io.WriteString(w, s)
-		if err != nil {
-			return n + int64(nn), err
-		}
-		n += int64(nn)
-	}
-	nn, err = io.WriteString(w, "}")
-	return n + int64(nn), err
-}
-
-// AppendTo appends the context in a canonical from to dst.
-//
-// AppendTo sorts the context keys and writes the sorted
-// key-value pairs as canonical JSON object to w.
-// Sort order is based on the un-escaped keys.
-//
-// Note that neither keys nor values are escaped for JSON.
-func (c Context) AppendTo(dst []byte) (output []byte) {
+// MarshalText sorts the context keys and writes the sorted
+// key-value pairs as canonical JSON object. The sort order
+// is based on the un-escaped keys.
+func (c Context) MarshalText() ([]byte, error) {
```

Comment:
Append is always more flexible than MarshalText. With append you can serialize directly into a byte stream without forcing the user to always copy the content.

You are making it less useful by removing this functionality. I you want to support the interface, wrapping is trivial.

```Go
func (c Context) MarshalText() ([]byte, error) {
	return c.AppendTo(make([]byte, 0, 128)), nil
}
```

As you can see it is forced to be up to 4x slower.

Reply:
That is one option. In any case, my reasoning is that while (obviously) appending to a pre-allocated buffer is cheaper w.r.t. memory allocs, the benchmark artificially shows a larger difference. It pre-allocs a buffer and appends the context multiple times - which is not how Context is used.

I wanted to support the text marshalling interface for the following reason:
```go
b, err := json.Marshal(Context{
   "key": "value",
   "key2": "value",
})
```
This may produce incorrect results as long as `Context` does not implement `MarshalJSON` or `MarshalText`.
As of now, there is no code that serializes the `Context` into a byte stream. Instead, it is always encoded as JSON.

Reply:
@aead As noted, a simply one line function makes it support MarshalText, so there really is no reason to rewrite the code to strings.Builder - it also uses a byte slice internally.

Reply:
@aead We can bike-shed this forever. My point is that there is no reason to implement an inferior algo, when a tested, existing one can satisfy what you want (support the interface) with a one-line wrapper, and we can avoid the error check in all uses of it.

I'll approve because I don't care enough. Do what yo want.